### PR TITLE
Add conda check and enable/disable options in plugin manager

### LINKED ISSUE #16: Plugin Manager shows conda as option for installation source in non-conda environment
## 🐛 Bug

As noted in napari/napari#5622 but deemed out of scope, indeed when running napari from a venv-based virtualenv (and all CONDA env vars removed) the Plugin Manager still shows conda as an option for plugin source in the Installation Info dropdown:

Screenshot 2023-03-15 at 10 20 01 AM

If I try to install a plugin using conda anyway, I get `ValueError: Prefix has not been specified!`.
<details><summary>Full stack trace</summary>

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/src/napari/napari/_qt/dialogs/qt_plugin_dialog.py:593, in QPluginList.handle_action(self=<napari._qt.dialogs.qt_plugin_dialog.QPluginList object>, item=<PyQt6.QtWidgets.QListWidgetItem object>, pkg_name='affinder==0.2.3', action_name='install', version='0.2.3', installer_choice='PyPI')
    588     pkg_name += (
    589         f"=={item.widget.version_choice_dropdown.currentText()}"
    590     )
    591 widget.set_busy(trans._("installing..."), action_name)
--> 593 job_id = self.installer.install(
        self = <napari._qt.dialogs.qt_plugin_dialog.QPluginList object at 0x2c970e830>
        self.installer = <napari._qt.dialogs.qt_package_installer.InstallerQueue object at 0x297c44550>
        tool = <InstallerTools.CONDA: 'conda'>
        pkg_name = 'affinder==0.2.3'
        [pkg_name] = ['affinder==0.2.3']
    594     tool=tool,
    595     pkgs=[pkg_name],
    596     # origins="TODO",
    597 )
    598 if self._warn_dialog:
    599     self._warn_dialog.exec_()

File ~/src/napari/napari/_qt/dialogs/qt_package_installer.py:311, in InstallerQueue.install(self=<napari._qt.dialogs.qt_package_installer.InstallerQueue object>, tool=<InstallerTools.CONDA: 'conda'>, pkgs=['affinder==0.2.3'], prefix=None, origins=(), **kwargs={})
    284 """Install packages in `pkgs` into `prefix` using `tool` with additional
    285 `origins` as source for `pkgs`.
    286 
   (...)
    301     ID that can be used to cancel the process.
    302 """
    303 item = self._build_queue_item(
    304     tool=tool,
    305     action=InstallerActions.INSTALL,
   (...)
    309     **kwargs,
    310 )
--> 311 return self._queue_item(item)
        item = CondaInstallerTool(action=<InstallerActions.INSTALL: 'install'>, pkgs=['affinder==0.2.3'], origins=(), prefix=None)
        self = <napari._qt.dialogs.qt_package_installer.InstallerQueue object at 0x297c44550>

File ~/src/napari/napari/_qt/dialogs/qt_package_installer.py:466, in InstallerQueue._queue_item(self=<napari._qt.dialogs.qt_package_installer.InstallerQueue object>, item=CondaInstallerTool(action=<InstallerActions.INST...kgs=['affinder==0.2.3'], origins=(), prefix=None))
    464 def _queue_item(self, item: AbstractInstallerTool) -> JobId:
    465     self._queue.append(item)
--> 466     self._process_queue()
        self = <napari._qt.dialogs.qt_package_installer.InstallerQueue object at 0x297c44550>
    467     return item.ident

File ~/src/napari/napari/_qt/dialogs/qt_package_installer.py:476, in InstallerQueue._process_queue(self=<napari._qt.dialogs.qt_package_installer.InstallerQueue object>)
    474 self.setProgram(str(tool.executable()))
    475 self.setProcessEnvironment(tool.environment())
--> 476 self.setArguments([str(arg) for arg in tool.arguments()])
        tool = CondaInstallerTool(action=<InstallerActions.INSTALL: 'install'>, pkgs=['affinder==0.2.3'], origins=(), prefix=None)
        self = <napari._qt.dialogs.qt_package_installer.InstallerQueue object at 0x297c44550>
    477 # this might throw a warning because the same process
    478 # was already running but it's ok
    479 self._log(
    480     trans._(
    481         "Starting '{program}' with args {args}",
   (...)
    484     )
    485 )

File ~/src/napari/napari/_qt/dialogs/qt_package_installer.py:189, in CondaInstallerTool.arguments(self=CondaInstallerTool(action=<InstallerActions.INST...kgs=['affinder==0.2.3'], origins=(), prefix=None))
    188 def arguments(self) -> Tuple[str, ...]:
--> 189     prefix = self.prefix or self._default_prefix()
        self.prefix = None
        self = CondaInstallerTool(action=<InstallerActions.INSTALL: 'install'>, pkgs=['affinder==0.2.3'], origins=(), prefix=None)
    190     if self.action == InstallerActions.UPGRADE:
    191         args = ['update', '-y', '--prefix', prefix]

File ~/src/napari/napari/_qt/dialogs/qt_package_installer.py:252, in CondaInstallerTool._default_prefix(self=CondaInstallerTool(action=<InstallerActions.INST...kgs=['affinder==0.2.3'], origins=(), prefix=None))
    250 if (Path(sys.prefix) / "conda-meta").is_dir():
    251     return sys.prefix
--> 252 raise ValueError("Prefix has not been specified!")

ValueError: Prefix has not been specified!
```
</details>

            
## To Reproduce

1. Create and activate a non-conda virtual environment (e.g. `python -m venv venv && source venv/bin/activate`)
2. Install napari in the venv (e.g. `pip install 'napari[pyqt6]'`, note I had to use qt6 because I'm on an ARM Macbook)
3. Launch napari, open the Plugin Manager from the "Plugins" menu, click the "Installation Info" dropdown under the `affinder` plugin, check the "Source" dropdown to see it offers both PyPI and conda.

## Expected behavior

I expect the Plugin Manager to not offer and not attempt to install using conda unless in a conda-based environment.

## Environment

```
napari: 0.4.17rc4.dev210+gc43a2e7c0.d20230302
Platform: macOS-13.2.1-arm64-arm-64bit
System: MacOS 13.2.1
Python: 3.10.9 (main, Feb 3 2023, 15:40:08) [Clang 14.0.0 (clang-1400.0.29.202)]
Qt: 6.4.2
PyQt6: 
NumPy: 1.24.1
SciPy: 1.10.0
Dask: 2023.1.1
VisPy: 0.12.2.dev38+dirty
magicgui: 0.6.1
superqt: unknown
in-n-out: 0.1.6
app-model: 0.1.1
npe2: 0.6.2

OpenGL:
- GL version: 2.1 Metal - 83
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 2560x1440, scale 1.0

Settings path:
- /Users/aandersoniii/Library/Application Support/napari/venv_640f7def1935afdf07a142187e645430c6d70fe6/settings.yaml
```

## Additional context
Note I don't have an entirely conda-free computer to test on, but I would still expect the behavior to be different in the setup I have here.

Summary of prior discussion - NB @Czaki @psobolewskiPhD:

 > > And that's where the widget is made. As far as I can tell there is no checks for when the widget shows conda. Unless I've missed something, I'd expect the widget to show conda regardless of how the env was created.
> > If conda isn't actually available then I think the only checks are here:
> 
> So it looks like there is a bug in the logic. Maybe out of the scope of this PR but at least should have issue about this.

> I think for sure out of scope, this is a small bug fix for a rather niche case.
If indeed conda is offered when not available or offered when non-condo env is used, that's a bigger deal I think. But best someone without conda test this first to make sure?

_Originally posted by @psobolewskiPhD in https://github.com/napari/napari/issues/5622#issuecomment-1468978782_

### PR DESCRIPTION
**Pull request by goanpeca**
# Description

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

Fixes https://github.com/napari/napari-plugin-manager/issues/16

# How has this been tested?

- [x] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change
- [x] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by goanpeca** [comment]
@aganders3 and @psobolewskiPhD, this should handle the cases you outlined. A third case I mentioned has to be dealt differently and will think about it 🤔 and open a separate PR.

**Comment by aganders3** [comment]
Testing in my non-conda env this works and I no longer get the option to install plugins with conda.

However, because of the issue you brought up with editable installs I'm actually not sure how to test this in a conda env now... My conda env with an editable install of napari also hides the conda option in the plugin manager.

**Comment by goanpeca** [comment]
>However, because of the issue you brought up with editable installs I'm actually not sure how to test this in a conda env now... My conda env with an editable install of napari also hides the conda option in the plugin manager.

Yes, not sure :-\

**Comment by goanpeca** [comment]
>However, because of the issue you brought up with editable installs I'm actually not sure how to test this in a conda env now... My conda env with an editable install of napari also hides the conda option in the plugin manager.

Well to test that it does work we could just add a dummy file called 

`napari-0.4.17devwhatever-0.json` in the `conda-meta` directory of the env, and then that should be enough to pass the `is_conda_package` check ?

**Comment by aganders3** [comment]
I built a conda package from this branch and installed it locally. The good news is it does show the conda option in the plugin manager. Unfortunately it's also now giving me the conda option for non-conda packages (e.g. `napari-boids`) with an empty list of versions.

I also get errors when attempting to install. I think the first error here is basically that `napari-boids` does not have a conda package. The second is maybe related to the fact that I built this napari conda package locally and it's trying to pin it when installing `natari`?
```
Starting '/Users/aandersoniii/mf/bin/conda' with args ['install', '-y', '--prefix', '/Users/aandersoniii/mf/envs/napari', '--override-channels', '-c', 'conda-forge', 'napari-boids==']

CondaValueError: invalid package specification: napari-boids==


Task finished with exit code 1 with status 0.
Starting '/Users/aandersoniii/mf/bin/conda' with args ['install', '-y', '--prefix', '/Users/aandersoniii/mf/envs/napari', '--override-channels', '-c', 'conda-forge', 'natari==0.2.7']
Collecting package metadata (current_repodata.json): ...working... 
done
Solving environment: ...working... 
failed with initial frozen solve. Retrying with flexible solve.
Solving environment: ...working... 
failed with repodata from current_repodata.json, will retry with next repodata source.


ResolvePackageNotFound: 
  - napari=0.5


Task finished with exit code 1 with status 0.
```

**Comment by goanpeca** [comment]
Hmmm I just pushed what I think should fix it, could you test again?

File: napari/_qt/dialogs/qt_plugin_dialog.py [hunk 1]
```diff
@@ -326,10 +307,10 @@ def setup_ui(self, enabled=True):
         self.version_choice_text = QLabel('Version:')
         self.source_choice_dropdown = QComboBox()
 
-        if len(self._versions_pypi) is not None:
+        if len(self._versions_pypi):
             self.source_choice_dropdown.addItem(PYPI)
 
-        if len(self._versions_conda) is not None:
+        if _IS_NAPARI_CONDA_INSTALLED and len(self._versions_conda):
```
Comment (Czaki) [inline napari/_qt/dialogs/qt_plugin_dialog.py]: Hm Reading this logic. 

call `len(None)` ends with TypeError. But the empty list evals to `False` and `None` evals to false so it should be changed to `if self._versions_pypi` and `if _IS_NAPARI_CONDA_INSTALLED and self._versions_conda`

File: napari/utils/_tests/test_conda.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+from napari.utils.conda import is_conda_package
+
+
+@pytest.mark.parametrize(
+    "pkg_name,expected",
+    [
+        ("some-package", True),
+        ("some-other-package", False),
+    ],
```
Comment (Czaki) [inline napari/utils/_tests/test_conda.py]:
```suggestion
    [
        ("some-package", True),
        ("some-other-package", False),
        ("some-package-other", False),
        ("other-some-package", False),
        ("package", False),
        ("some", False),
    ],
```

Use prefixed/suffixed names of package

File: napari/_qt/dialogs/qt_plugin_dialog.py (hunk 1 above)
Comment (goanpeca) [reply napari/_qt/dialogs/qt_plugin_dialog.py]: Fixed!

File: napari/utils/conda.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    Returns
+    -------
+    bool: True if a conda package, False if not
+
+    Notes
+    -----
+    Installed conda packages within a conda installation and environment can
+    be identified as `<package-name>-<version>-<build-string>.json` files
+    saved within a `conda-meta` folder within the given environment of
```
Comment (Carreau) [inline napari/utils/conda.py]:
```suggestion
    be identified as ``<package-name>-<version>-<build-string>.json`` files
    saved within a ``conda-meta`` folder within the given environment of
```

Not markdown :-)

**Comment by aganders3** [comment]
This now appears to be working correctly for me, with the exception of [what is already noted as FIXME](https://github.com/goanpeca/napari/blob/84fa8b2b26374c766f8575bc76ebaf6afc818e82/napari/_qt/dialogs/qt_package_installer.py#L219). So I can't actually test installing conda packages, but testing both conda- and venv-based virtual environments shows what I would expect for the plugin installation sources.

**Comment by goanpeca** [comment]
Adding the ready to merge label and start counter @napari/core-devs ? 🚀

**Review by Czaki — APPROVED** [review]

**Review by psobolewskiPhD — DISMISSED** [review]
So i had a chance to test this. I used 2 envs, one where I have **everything** from pip (pyside6, arm64). There I had just pip offered in the the installation options and it worked fine (install and upgrade).

I also made an env with the conda nightly, so everything was from conda.
There I had both conda and pip offered. Using conda worked nicely and if I installed napari-animation 0.0.3 it would update also using conda.

**However** using the default (pip), for example for napari-animation though gave
```
ERROR: Cannot install napari-animation==0.0.5 because these package versions have conflicting dependencies.

The conflict is caused by:
    napari-animation 0.0.5 depends on napari
    The user requested (constraint) napari==0.5.0a2.dev55+g8a793b9f
```
I don't get this because napari-animation doesn't pin a napari version and pip from the command line happily installs with:
```
Requirement already satisfied: napari in /Users/piotrsobolewski/Dev/miniforge3/envs/napari-main/lib/python3.10/site-packages (from napari-animation) (0.5.0a2.dev55+g8a793b9f)
```

This is important, because **pip** is the default and many (most?) plugins list napari as a dependency and as a resultdon't install. 
Now if we want to prevent pip install in pure conda env—not unreasonable—then we should explicitly do that, or make conda the default in that case.

Anyhow, I'd like a clarification.

**Comment by aganders3** [comment]
I _think_ this constraint comes from trying to pin napari to the _current_ installed version, such that installing a plugin that does depend on napari does not accidentally update your installed version of napari (a previous bug I think maybe you called out, @psobolewskiPhD?).

How did you install napari in the test environment? I think this is similar to the issue I had (in reverse) above and is called out in the code here: https://github.com/goanpeca/napari/blob/84fa8b2b26374c766f8575bc76ebaf6afc818e82/napari/_qt/dialogs/qt_package_installer.py#L219

All this to say I don't have a great solution in mind here 😕. I can give it some thought though.

**Comment by psobolewskiPhD** [comment]
In installed conda nightly napari using:
`conda create -n napari-main -c "napari/label/nightly" -c conda-forge napari python=3.10`



I did raise the issue of napari getting downgraded by a plugin that had a napari version pin, which happens with a regular pip install as well. So here, with the manager, If a plugin pinned 0.4.17 (or lower) and I have 0.5.0dev I'd expect the install to fail. 

But regular pip install seems content with a *newer napari* being present if a plugin depends on just `napari`. So it's a surprise to me that this pip install via the manager does not. I hope this is an edge case of dev napari, but it's a real use case: the conda nightly is an easy way to use napari given how long it's been since we had a release, so installing plugins by default method should work.

And yeah I saw that FIXME, but that's in the conda installer code? while I'm using pip install (the default), so I think that's using:
https://github.com/goanpeca/napari/blob/84fa8b2b26374c766f8575bc76ebaf6afc818e82/napari/_qt/dialogs/qt_package_installer.py#L89

**Comment by aganders3** [comment]
> I hope this is an edge case of dev napari, but it's a real use case: the conda nightly is an easy way to use napari given how long it's been since we had a release, so installing plugins by default method should work.

💯 - we should make this work

**Comment by Czaki** [comment]
It will be nice to have information if a plugin has napari in its dependencies. As if the napari is not in dependencies or is in optional dependencies and conda env is used then the plugin could be installed from conda.

**Comment by psobolewskiPhD** [comment]
Naturally, I'm an idiot, so using the conda nightly doesn't test this PR 😬 
@aganders3 Looks like a difference in behavior between conda and mamba.
Starting with:
image
Mamba succeeds with no errors:
```
Starting '/Users/piotrsobolewski/Dev/miniforge3/bin/mamba' with args ['update', '-y', '--prefix', '/Users/piotrsobolewski/Dev/miniforge3/envs/napari-main', '--override-channels', '-c', 'conda-forge', 'napari-animation==0.0.5']
```
<details>

```
conda-forge/osx-arm64                                       Using cache
conda-forge/noarch                                          Using cache

Transaction

  Prefix: /Users/piotrsobolewski/Dev/miniforge3/envs/napari-main

  Updating specs:

   - napari-animation==0.0.5

   - ca-certificates
   - certifi
   - openssl

  Package             Version  Build         Channel                  Size
────────────────────────────────────────────────────────────────────────────
  Upgrade:
────────────────────────────────────────────────────────────────────────────

  - napari-animation    0.0.3  pyhd8ed1ab_0  conda-forge                  
  + napari-animation    0.0.5  pyhd8ed1ab_0  conda-forge/noarch     Cached

  Summary:
  Upgrade: 1 packages
  Total download: 0 B

────────────────────────────────────────────────────────────────────────────

Looking for: ['napari-animation==0.0.5']

Pinned packages:
  - python 3.10.*
  - napari 0.5.*
  - 
Preparing transaction: ...working... 
done
Verifying transaction: ...working... 
done
Executing transaction: ...working... 
done
Task finished with exit code 0 with status 0.
```

</details>

And that's despite the napari 0.5.*  too... 🤔 It did work: 
```
napari-animation          0.0.5              pyhd8ed1ab_0    conda-forge
```
So the update behavior needs to be fixed. The conda docs indicate that `conda update` installs the latest version, rather than a specific one like that commands wants.

And yes with this conda nightly (again, not this PR), pip install (the default) does work.

I made a new pip install env and tested things there and also pip install was OK, so I guess I had something messed up with envs or git branches?

So my action item is to dismiss the erroneous review, but we should double check the `conda update` behavior!

**Review by psobolewskiPhD — CHANGES_REQUESTED** [review]
Hi, disregard my other review, it was in error. But @aganders3 did find an issue, that upgrades don't work with **conda** (fine with mamba), see:
https://github.com/napari/napari/pull/5641#issuecomment-1494567176

I think the source of the issue is here:
https://github.com/goanpeca/napari/blob/33156bf839ead78a79776c9288888cbbfa1470b8/napari/_qt/dialogs/qt_plugin_dialog.py#L584-L586
```
        if action_name == InstallerActions.UPGRADE:
            if hasattr(item, 'latest_version'):
                pkg_name += f"=={item.latest_version}"
```
the pkg_name is being set to `pkg_name==latest_version`
conda doesn't like. The upgrade button at the top aims to upgrade to the latest, so no version should be passed to conda/mamba and the latest will be used anyways.
And the same for pip upgrades either right? pip uses `pip install -U package` to install  latest.
So I think appending the ==<version> can be removed.

**Comment by kephale** [comment]
From the PR party, @goanpeca should we migrate this to https://github.com/napari/napari-plugin-manager ?

**Comment by psobolewskiPhD** [comment]
@goanpeca we discussed this at the PR party.
Given the split off of the plugin manager and the fact that that won't `go live` till 0.5.0, then this PR should probably just me migrated to the other repo. 
What do you think?

**Comment by goanpeca** [comment]
>@goanpeca we discussed this at the PR party.
>Given the split off of the plugin manager and the fact that that won't go live till 0.5.0, then this PR should probably just me migrated to the other repo.
>What do you think?

Sounds good, will do it!

**Pull request closed without merging**